### DISCUSSION TITLE
Preserve blame in squash merge

### DISCUSSION THREAD

**Discussion by anirban-ray-publicis-sapient** [community/community#38240] [New to GitHub]
I like to create a lot of small commits while working on a feature branch for every sub-items. So I usually choose to squash them while creating a merge request to keep a smaller history.

However, I observed that this leads to loss of the intermediate commits. If I see the blame, every change shows the squashing commit and not of the original one. And also, if it is merged by someone else, the author of these lines vet changed to that person.

I’d prefer to have this. When the squashing merge happens, the target branch will get just one more commit. However, blame will be as it is with non-squashing merge.

Is there a way to achieve this?

**Comment by ossanna16** [comment]
Hi there @anirban-ray-publicis-sapient and welcome to our community! Thank you for asking a great question 🙂
While you are waiting for a reply to your question, here are a few things you can do:

To get started, introduce yourself in our [official introduction thread](https://github.com/orgs/community/discussions/23204)
To connect with other new users, check out this Discussion to [Meet Your GitHub Community Team](https://github.com/orgs/community/discussions/28776), further your GitHub knowledge by working through our [GitHub Skill 🆙 exercises](https://github.com/orgs/community/discussions/32121), and participate in our [weekly Fun Friday chats](https://github.com/orgs/community/discussions/30912)

**Comment by airtower-luna** [comment]
You can't have both: A squash "merge" doesn't really merge the branches, it creates a new commit that contains the sum of the changes the head branch of the PR adds, and *discards* the original commits. If you want to preserve the information from the original commits, you need to do a regular merge (or a local ff-merge, if possible). A rebase "merge" mostly works, too: It preserves author information, but not the original commit IDs.

**Reply by anirban-ray-publicis-sapient** [reply]
Hi @airtower-luna, thanks for the answer. I really want to keep the original commit message/author information, and simultaneously want to minimise the history of main branch.

I haven't used "rebase merge" before. I thought rebase  works similarly as `reset --soft HEAD~k` + `stash push` + `merge` + `stash pop` type of way, but that is unlikely to solve my case. I have probably misunderstood it, can you please give an example?

**Reply by airtower-luna** [reply]
The rebase "merge" (not really a merge) recreates the commits on the head branch on top of the target branch, avoiding the merge commit, and keeping the commit structure. That's the closest you can get to a fast-forward merge if you want to merge via the web UI.

I'll reiterate that the thing you're asking for (having just one commit with the commit information from N previous commits) is simply not possible. All you can avoid is having a separate merge commit.
